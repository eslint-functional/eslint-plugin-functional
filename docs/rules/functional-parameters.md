# Enforce Functional Parameters (functional-parameters)

Disallow use of rest parameters, the `arguments` keyword and enforces that functions take a least 1 parameter.

## Rule Details

In functions, `arguments` is a special variable that is implicitly available.
This variable is an array containing the arguments passed to the function call and is often used to allow for any number of parameters to be passed to a function. Rest parameters are another way any number of parameters can be passed to a function.

When it comes to functional programming, known and explicit parameters must be used.

Note: With an unknown number of parameters, currying functions is a lot more difficult/impossible.

Examples of **incorrect** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add() {
  return arguments.reduce((sum, number) => sum + number, 0);
}
```

```js
/* eslint functional/functional-parameters: "error" */

function add(...numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

Examples of **correct** code for this rule:

```js
/* eslint functional/functional-parameters: "error" */

function add(numbers) {
  return numbers.reduce((sum, number) => sum + number, 0);
}
```

## Options

This rule accepts an options object of the following type:

```ts
{
  allowRestParameter: boolean;
  allowArgumentsKeyword: boolean;
  enforceParameterCount: false | "atLeastOne" | "exactlyOne" | {
    count: "atLeastOne" | "exactlyOne";
    ignoreIIFE: boolean;
  };
  ignorePattern?: string | Array<string>;
}
```

The default options:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: {
    count: "atLeastOne",
    ignoreIIFE: true
  }
}
```

Note: the `lite` ruleset overrides the default options to:

```ts
{
  allowRestParameter: false,
  allowArgumentsKeyword: false,
  enforceParameterCount: false
}
```

### `allowRestParameter`

If true, this option allows for the use of rest parameters.

### `allowArgumentsKeyword`

If true, this option allows for the use of the `arguments` keyword.

### `enforceParameterCount`

This option enforces the number of parameters each function takes.

#### false

Don't enforce a parameter count.

#### "atLeastOne"

Require all functions to have a least one parameter.

There's not much point of having a function that doesn't take any parameters in functional programming.

#### "exactlyOne"

Require all functions to have exactly one parameter.

Any function that take takes multiple parameter can be rewritten as a higher-order function that only takes one.

Example:

```js
// This function
function add(x, y) {
  return x + y;
}

// can be rewritten like this.
function add(x) {
  return (y) => x + y;
}
```

See [Currying](https://en.wikipedia.org/wiki/Currying) and [Higher-order function](https://en.wikipedia.org/wiki/Higher-order_function) on Wikipedia for more infomation.

#### `enforceParameterCount.count`

See [enforceParameterCount](#enforceparametercount).

#### `enforceParameterCount.ignoreIIFE`

If true, this option allows for the use of [IIFEs](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) that do not have any parameters.

### `ignorePattern`

Patterns will be matched against function names.
See the [ignorePattern](./options/ignore-pattern.md) docs for more infomation.
