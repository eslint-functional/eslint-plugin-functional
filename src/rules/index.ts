import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClass from "./no-class";
import * as noConditionalStatement from "./no-conditional-statement";
import * as noExpressionStatement from "./no-expression-statement";
import * as noLet from "./no-let";
import * as noLoop from "./no-loop-statement";
import * as noMethodSignature from "./no-method-signature";
import * as noMixedType from "./no-mixed-type";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpression from "./no-this-expression";
import * as noThrowStatement from "./no-throw-statement";
import * as noTryStatement from "./no-try-statement";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferType from "./prefer-type-literal";

/**
 * All of the custom rules.
 */
export const rules = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClass.name]: noClass.rule,
  [noConditionalStatement.name]: noConditionalStatement.rule,
  [noExpressionStatement.name]: noExpressionStatement.rule,
  [noLet.name]: noLet.rule,
  [noLoop.name]: noLoop.rule,
  [noMethodSignature.name]: noMethodSignature.rule,
  [noMixedType.name]: noMixedType.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpression.name]: noThisExpression.rule,
  [noThrowStatement.name]: noThrowStatement.rule,
  [noTryStatement.name]: noTryStatement.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferType.name]: preferType.rule,
};
