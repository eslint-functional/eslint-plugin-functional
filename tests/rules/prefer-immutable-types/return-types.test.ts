import dedent from "dedent";
import { createRuleTester } from "eslint-vitest-rule-tester";
import { describe, expect, it } from "vitest";

import { name, rule } from "#/rules/prefer-immutable-types";

import { typescriptConfig } from "../../utils/configs";

describe(name, () => {
  describe("typescript", () => {
    const { valid, invalid } = createRuleTester({
      name,
      rule,
      configs: typescriptConfig,
    });

    it("reports non-immutable set return types", async () => {
      const invalidResult = await invalid({
        code: "function foo(): ReadonlySet<string> {}",
        errors: ["returnType"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports non-immutable map return types", async () => {
      const invalidResult = await invalid({
        code: "function foo(): ReadonlyMap<string, string> {}",
        errors: ["returnType"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports mutable records return types", async () => {
      const invalidResult = await invalid({
        code: "function foo(): { foo: string } {}",
        errors: ["returnType"],
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it("reports mutable records return types and suggests a fix for ReadonlyShallow", async () => {
      const invalidResult = await invalid({
        code: "function foo(): { foo: string } {}",
        options: [{ returnTypes: "ReadonlyShallow" }],
        errors: ["returnType"],
      });
      expect(invalidResult.result).toMatchSnapshot();
      for (const message of invalidResult.result.messages) {
        expect(message.suggestions).toHaveLength(1);
      }
    });

    it("doesn't report type assertion return types", async () => {
      await valid({
        code: "function foo(arg0: { foo: string | number }, arg1: { foo: string | number }): arg0 is { foo: number } {}",
        options: [
          {
            parameters: false,
            returnTypes: "ReadonlyShallow",
          },
        ],
      });
    });

    it("suggest multiple fixes for collections for ReadonlyShallow", async () => {
      const invalidResult = await invalid({
        code: dedent`
          function foo(): Array<string> {}
          function foo(): string[] {}
          function foo(): Set<string> {}
          function foo(): Map<string, string> {}
          function foo(): ReadonlyArray<string> {}
          function foo(): readonly string[] {}
          function foo(): ReadonlySet<string> {}
          function foo(): ReadonlyMap<string, string> {}
        `,
        options: [{ returnTypes: "ReadonlyShallow" }],
        errors: ["returnType", "returnType", "returnType", "returnType"],
      });
      expect(invalidResult.result).toMatchSnapshot();
      for (const message of invalidResult.result.messages) {
        expect(message.suggestions).toHaveLength(2);
      }
    });

    it("allows for user suggestions", async () => {
      const invalidResult = await invalid({
        code: "function foo(): { foo: string } {}",
        options: [
          {
            returnTypes: "ReadonlyDeep",
            suggestions: {
              ReadonlyDeep: [
                [
                  {
                    pattern: "^(?!ReadonlyDeep)(?:Readonly<(.+)>|(.+))$",
                    replace: "ReadonlyDeep<$1$2>",
                  },
                ],
              ],
            },
          },
        ],
        errors: ["returnType"],
      });
      expect(invalidResult.result).toMatchSnapshot();
      for (const message of invalidResult.result.messages) {
        expect(message.suggestions).toHaveLength(1);
      }
    });

    it("allows for user fixes", async () => {
      const invalidResult = await invalid({
        code: dedent`
          function foo(): Array<string> {}
          function foo(): string[] {}
          function foo(): Set<string> {}
          function foo(): Map<string, string> {}
          function foo(): ReadonlyArray<string> {}
          function foo(): readonly string[] {}
          function foo(): ReadonlySet<string> {}
          function foo(): ReadonlyMap<string, string> {}
        `,
        options: [
          {
            returnTypes: "Immutable",
            fixer: {
              Immutable: [
                {
                  pattern: "^(?:Readonly)?(Set|Map)<(.+)>$",
                  replace: "Readonly<Readonly$1<$2>>",
                },
              ],
            },
          },
        ],
        errors: [
          "returnType",
          "returnType",
          "returnType",
          "returnType",
          "returnType",
          "returnType",
          "returnType",
          "returnType",
        ],
        verifyAfterFix: false, // "fix" doesn't fix arrays so they will still report.
      });
      expect(invalidResult.result).toMatchSnapshot();
    });

    it.each([
      [[{ returnTypes: "ReadonlyShallow" }]],
      [[{ returnTypes: "ReadonlyDeep" }]],
      [[{ returnTypes: "Immutable" }]],
    ])("doesn't reports valid return types", async (options) => {
      await valid({
        code: "function foo(): boolean {}",
        options,
      });
      await valid({
        code: "function foo(): true {}",
        options,
      });
      await valid({
        code: "function foo(): string {}",
        options,
      });
      await valid({
        code: "function foo(): 'bar' {}",
        options,
      });
      await valid({
        code: "function foo(): undefined {}",
        options,
      });
      await valid({
        code: "function foo(): readonly string[] {}",
        options,
        settings: {
          immutability: {
            overrides: [
              {
                type: { from: "lib", name: "ReadonlyArray" },
                to: "Immutable",
              },
            ],
          },
        },
      });
      await valid({
        code: "function foo(): { readonly foo: string } {}",
        options,
      });
      await valid({
        code: "function foo(): { readonly foo: { readonly bar: number } } {}",
        options,
      });
      await valid({
        code: "function foo(): Readonly<ReadonlySet<string>> {}",
        options,
      });
      await valid({
        code: "function foo(): Readonly<ReadonlyMap<string, string>> {}",
        options,
      });
      await valid({
        code: "function foo(arg: { foo: string | number }): arg is { foo: number } {}",
        options,
      });

      if (options[0]!.returnTypes !== "Immutable") {
        await valid({
          code: "function foo(): { foo(): void } {}",
          options,
        });
        await valid({
          code: "function foo(): ReadonlyArray<string> {}",
          options,
        });
        await valid({
          code: "function foo(): readonly [string, number] {}",
          options,
        });
        await valid({
          code: "function foo(): Readonly<[string, number]> {}",
          options,
        });
        await valid({
          code: "function foo(): { foo: () => void } {}",
          options,
        });
        await valid({
          code: "function foo(): ReadonlySet<string> {}",
          options,
        });
        await valid({
          code: "function foo(): ReadonlyMap<string, string> {}",
          options,
        });
        await valid({
          code: dedent`
            interface Foo {
              (): readonly string[];
            }
          `,
          options,
        });
        await valid({
          code: dedent`
            interface Foo {
              new (): readonly string[];
            }
          `,
          options,
        });
      }
    });
  });
});
