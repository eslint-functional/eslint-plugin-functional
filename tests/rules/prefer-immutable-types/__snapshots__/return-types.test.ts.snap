// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefer-immutable-types > typescript > allows for user fixes 1`] = `
{
  "fixed": true,
  "messages": [
    {
      "column": 15,
      "endColumn": 30,
      "endLine": 1,
      "line": 1,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 15,
      "endColumn": 25,
      "endLine": 2,
      "line": 2,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 15,
      "endColumn": 28,
      "endLine": 3,
      "fix": {
        "range": [
          77,
          88,
        ],
        "text": "Readonly<ReadonlySet<string>>",
      },
      "line": 3,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 15,
      "endColumn": 36,
      "endLine": 4,
      "fix": {
        "range": [
          108,
          127,
        ],
        "text": "Readonly<ReadonlyMap<string, string>>",
      },
      "line": 4,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 15,
      "endColumn": 38,
      "endLine": 5,
      "line": 5,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 15,
      "endColumn": 34,
      "endLine": 6,
      "line": 6,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 15,
      "endColumn": 36,
      "endLine": 7,
      "fix": {
        "range": [
          225,
          244,
        ],
        "text": "Readonly<ReadonlySet<string>>",
      },
      "line": 7,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
    {
      "column": 15,
      "endColumn": 44,
      "endLine": 8,
      "fix": {
        "range": [
          264,
          291,
        ],
        "text": "Readonly<ReadonlyMap<string, string>>",
      },
      "line": 8,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
  ],
  "output": "function foo(): Array<string> {}
function foo(): string[] {}
function foo(): Readonly<ReadonlySet<string>> {}
function foo(): Readonly<ReadonlyMap<string, string>> {}
function foo(): ReadonlyArray<string> {}
function foo(): readonly string[] {}
function foo(): Readonly<ReadonlySet<string>> {}
function foo(): Readonly<ReadonlyMap<string, string>> {}",
  "steps": [
    {
      "fixed": true,
      "messages": [
        {
          "column": 15,
          "endColumn": 30,
          "endLine": 1,
          "line": 1,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 15,
          "endColumn": 25,
          "endLine": 2,
          "line": 2,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 15,
          "endColumn": 38,
          "endLine": 5,
          "line": 5,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 15,
          "endColumn": 34,
          "endLine": 6,
          "line": 6,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "function foo(): Array<string> {}
function foo(): string[] {}
function foo(): Readonly<ReadonlySet<string>> {}
function foo(): Readonly<ReadonlyMap<string, string>> {}
function foo(): ReadonlyArray<string> {}
function foo(): readonly string[] {}
function foo(): Readonly<ReadonlySet<string>> {}
function foo(): Readonly<ReadonlyMap<string, string>> {}",
    },
    {
      "fixed": false,
      "messages": [
        {
          "column": 15,
          "endColumn": 30,
          "endLine": 1,
          "line": 1,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 15,
          "endColumn": 25,
          "endLine": 2,
          "line": 2,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 15,
          "endColumn": 38,
          "endLine": 5,
          "line": 5,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
        {
          "column": 15,
          "endColumn": 34,
          "endLine": 6,
          "line": 6,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "function foo(): Array<string> {}
function foo(): string[] {}
function foo(): Readonly<ReadonlySet<string>> {}
function foo(): Readonly<ReadonlyMap<string, string>> {}
function foo(): ReadonlyArray<string> {}
function foo(): readonly string[] {}
function foo(): Readonly<ReadonlySet<string>> {}
function foo(): Readonly<ReadonlyMap<string, string>> {}",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > allows for user suggestions 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 15,
      "endColumn": 32,
      "endLine": 1,
      "line": 1,
      "message": "Return type should have an immutability of at least "ReadonlyDeep" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Replace with: ReadonlyDeep<{ foo: string }>",
          },
          "desc": "Replace with: ReadonlyDeep<{ foo: string }>",
          "fix": {
            "range": [
              16,
              31,
            ],
            "text": "ReadonlyDeep<{ foo: string }>",
          },
          "messageId": "userDefined",
        },
      ],
    },
  ],
  "output": "function foo(): { foo: string } {}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 15,
          "endColumn": 32,
          "endLine": 1,
          "line": 1,
          "message": "Return type should have an immutability of at least "ReadonlyDeep" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Replace with: ReadonlyDeep<{ foo: string }>",
              },
              "desc": "Replace with: ReadonlyDeep<{ foo: string }>",
              "fix": {
                "range": [
                  16,
                  31,
                ],
                "text": "ReadonlyDeep<{ foo: string }>",
              },
              "messageId": "userDefined",
            },
          ],
        },
      ],
      "output": "function foo(): { foo: string } {}",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > reports mutable records return types 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 15,
      "endColumn": 32,
      "endLine": 1,
      "line": 1,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
  ],
  "output": "function foo(): { foo: string } {}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 15,
          "endColumn": 32,
          "endLine": 1,
          "line": 1,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "function foo(): { foo: string } {}",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > reports mutable records return types and suggests a fix for ReadonlyShallow 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 15,
      "endColumn": 32,
      "endLine": 1,
      "line": 1,
      "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              16,
              31,
            ],
            "text": "Readonly<{ foo: string }>",
          },
          "messageId": "userDefined",
        },
      ],
    },
  ],
  "output": "function foo(): { foo: string } {}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 15,
          "endColumn": 32,
          "endLine": 1,
          "line": 1,
          "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  16,
                  31,
                ],
                "text": "Readonly<{ foo: string }>",
              },
              "messageId": "userDefined",
            },
          ],
        },
      ],
      "output": "function foo(): { foo: string } {}",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > reports non-immutable map return types 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 15,
      "endColumn": 44,
      "endLine": 1,
      "line": 1,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
  ],
  "output": "function foo(): ReadonlyMap<string, string> {}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 15,
          "endColumn": 44,
          "endLine": 1,
          "line": 1,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "function foo(): ReadonlyMap<string, string> {}",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > reports non-immutable set return types 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 15,
      "endColumn": 36,
      "endLine": 1,
      "line": 1,
      "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
    },
  ],
  "output": "function foo(): ReadonlySet<string> {}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 15,
          "endColumn": 36,
          "endLine": 1,
          "line": 1,
          "message": "Return type should have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
        },
      ],
      "output": "function foo(): ReadonlySet<string> {}",
    },
  ],
}
`;

exports[`prefer-immutable-types > typescript > suggest multiple fixes for collections for ReadonlyShallow 1`] = `
{
  "fixed": false,
  "messages": [
    {
      "column": 15,
      "endColumn": 30,
      "endLine": 1,
      "line": 1,
      "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Use ReadonlyArray instead of Array.",
          },
          "desc": "Use ReadonlyArray instead of Array.",
          "fix": {
            "range": [
              16,
              29,
            ],
            "text": "ReadonlyArray<string>",
          },
          "messageId": "userDefined",
        },
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              16,
              29,
            ],
            "text": "Readonly<Array<string>>",
          },
          "messageId": "userDefined",
        },
      ],
    },
    {
      "column": 15,
      "endColumn": 25,
      "endLine": 2,
      "line": 2,
      "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Prepend with readonly.",
          },
          "desc": "Prepend with readonly.",
          "fix": {
            "range": [
              49,
              57,
            ],
            "text": "readonly string[]",
          },
          "messageId": "userDefined",
        },
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              49,
              57,
            ],
            "text": "Readonly<string[]>",
          },
          "messageId": "userDefined",
        },
      ],
    },
    {
      "column": 15,
      "endColumn": 28,
      "endLine": 3,
      "line": 3,
      "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Use ReadonlySet instead of Set.",
          },
          "desc": "Use ReadonlySet instead of Set.",
          "fix": {
            "range": [
              77,
              88,
            ],
            "text": "ReadonlySet<string>",
          },
          "messageId": "userDefined",
        },
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              77,
              88,
            ],
            "text": "Readonly<Set<string>>",
          },
          "messageId": "userDefined",
        },
      ],
    },
    {
      "column": 15,
      "endColumn": 36,
      "endLine": 4,
      "line": 4,
      "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
      "messageId": "returnType",
      "nodeType": "TSTypeAnnotation",
      "ruleId": "prefer-immutable-types",
      "severity": 2,
      "suggestions": [
        {
          "data": {
            "message": "Use ReadonlyMap instead of Map.",
          },
          "desc": "Use ReadonlyMap instead of Map.",
          "fix": {
            "range": [
              108,
              127,
            ],
            "text": "ReadonlyMap<string, string>",
          },
          "messageId": "userDefined",
        },
        {
          "data": {
            "message": "Surround with Readonly.",
          },
          "desc": "Surround with Readonly.",
          "fix": {
            "range": [
              108,
              127,
            ],
            "text": "Readonly<Map<string, string>>",
          },
          "messageId": "userDefined",
        },
      ],
    },
  ],
  "output": "function foo(): Array<string> {}
function foo(): string[] {}
function foo(): Set<string> {}
function foo(): Map<string, string> {}
function foo(): ReadonlyArray<string> {}
function foo(): readonly string[] {}
function foo(): ReadonlySet<string> {}
function foo(): ReadonlyMap<string, string> {}",
  "steps": [
    {
      "fixed": false,
      "messages": [
        {
          "column": 15,
          "endColumn": 30,
          "endLine": 1,
          "line": 1,
          "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Use ReadonlyArray instead of Array.",
              },
              "desc": "Use ReadonlyArray instead of Array.",
              "fix": {
                "range": [
                  16,
                  29,
                ],
                "text": "ReadonlyArray<string>",
              },
              "messageId": "userDefined",
            },
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  16,
                  29,
                ],
                "text": "Readonly<Array<string>>",
              },
              "messageId": "userDefined",
            },
          ],
        },
        {
          "column": 15,
          "endColumn": 25,
          "endLine": 2,
          "line": 2,
          "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Prepend with readonly.",
              },
              "desc": "Prepend with readonly.",
              "fix": {
                "range": [
                  49,
                  57,
                ],
                "text": "readonly string[]",
              },
              "messageId": "userDefined",
            },
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  49,
                  57,
                ],
                "text": "Readonly<string[]>",
              },
              "messageId": "userDefined",
            },
          ],
        },
        {
          "column": 15,
          "endColumn": 28,
          "endLine": 3,
          "line": 3,
          "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Use ReadonlySet instead of Set.",
              },
              "desc": "Use ReadonlySet instead of Set.",
              "fix": {
                "range": [
                  77,
                  88,
                ],
                "text": "ReadonlySet<string>",
              },
              "messageId": "userDefined",
            },
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  77,
                  88,
                ],
                "text": "Readonly<Set<string>>",
              },
              "messageId": "userDefined",
            },
          ],
        },
        {
          "column": 15,
          "endColumn": 36,
          "endLine": 4,
          "line": 4,
          "message": "Return type should have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
          "messageId": "returnType",
          "nodeType": "TSTypeAnnotation",
          "ruleId": "rule-to-test/prefer-immutable-types",
          "severity": 2,
          "suggestions": [
            {
              "data": {
                "message": "Use ReadonlyMap instead of Map.",
              },
              "desc": "Use ReadonlyMap instead of Map.",
              "fix": {
                "range": [
                  108,
                  127,
                ],
                "text": "ReadonlyMap<string, string>",
              },
              "messageId": "userDefined",
            },
            {
              "data": {
                "message": "Surround with Readonly.",
              },
              "desc": "Surround with Readonly.",
              "fix": {
                "range": [
                  108,
                  127,
                ],
                "text": "Readonly<Map<string, string>>",
              },
              "messageId": "userDefined",
            },
          ],
        },
      ],
      "output": "function foo(): Array<string> {}
function foo(): string[] {}
function foo(): Set<string> {}
function foo(): Map<string, string> {}
function foo(): ReadonlyArray<string> {}
function foo(): readonly string[] {}
function foo(): ReadonlySet<string> {}
function foo(): ReadonlyMap<string, string> {}",
    },
  ],
}
`;
